File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/content/JcrCommentManager.java
package solutions.digamma.damas.jcr.content;

import solutions.digamma.damas.common.CompatibilityException;
import solutions.digamma.damas.common.WorkspaceException;
import solutions.digamma.damas.entity.Page;
import solutions.digamma.damas.content.Comment;
import solutions.digamma.damas.content.CommentManager;
import solutions.digamma.damas.inspection.NotNull;
import solutions.digamma.damas.jcr.names.TypeNamespace;
import solutions.digamma.damas.jcr.model.JcrCrudManager;
import solutions.digamma.damas.jcr.model.JcrSearchEngine;

import javax.jcr.Node;
import javax.jcr.RepositoryException;
import javax.jcr.Session;
import java.util.UUID;

/**
 * JCR-based comment manager implementation.
 *
 * @author Ahmad Shahwan
 */
public class JcrCommentManager
        extends JcrCrudManager<Comment>
        implements JcrSearchEngine<Comment>, CommentManager {

    @Override
    protected JcrComment retrieve(Session session, String id)
            throws RepositoryException, WorkspaceException {
        return new JcrComment(session.getNodeByIdentifier(id));
    }

    @Override
    protected JcrComment create(Session session, @NotNull Comment entity)
            throws RepositoryException, WorkspaceException {
        String name = UUID.randomUUID().toString();
        Node parent = session.getNodeByIdentifier(entity.getReceiverId());
        if (!parent.isNodeType(TypeNamespace.COMMENT_RECEIVER)) {
            throw new CompatibilityException(
                    "Parent cannot receive comments.");
        }
        Node node = parent.addNode(name, TypeNamespace.COMMENT);
        JcrComment comment = new JcrComment(node);
        comment.update(entity);
        return comment;
    }

    @Override
    protected JcrComment update(
            Session session, String id,
            @NotNull Comment entity)
            throws RepositoryException, WorkspaceException {
        JcrComment comment = this.retrieve(session, id);
        comment.update(entity);
        return comment;
    }

    @Override
    protected void delete(Session session, String id)
            throws RepositoryException, WorkspaceException {
        JcrComment comment = this.retrieve(session, id);
        comment.remove();
    }

    @Override
    public Page<Comment> find(
            Session session,
            int offset,
            int size,
            Object query)
            throws RepositoryException, WorkspaceException {
        return null;
    }
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/content/JcrCommentManager.kt
package solutions.digamma.damas.jcr.content

import solutions.digamma.damas.common.WorkspaceException
import solutions.digamma.damas.content.Comment
import solutions.digamma.damas.content.CommentManager
import solutions.digamma.damas.jcr.model.JcrCrudManager
import solutions.digamma.damas.jcr.model.JcrSearchEngine
import solutions.digamma.damas.jcr.names.TypeNamespace
import javax.inject.Singleton
import javax.jcr.Node
import javax.jcr.RepositoryException

/**
 * JCR-based comment manager implementation.
 *
 * @author Ahmad Shahwan
 */
@Singleton
internal open class JcrCommentManager :
        JcrCrudManager<Comment>(), JcrSearchEngine<Comment>, CommentManager {

    @Throws(RepositoryException::class, WorkspaceException::class)
    override fun doRetrieve(id: String) =
        JcrComment.of(this.session.getNodeByIdentifier(id))

    @Throws(RepositoryException::class, WorkspaceException::class)
    override fun doCreate(pattern: Comment): JcrComment {
        return JcrComment.from(this.session, pattern.receiverId).also {
            it.update(pattern)
        }
    }

    @Throws(RepositoryException::class, WorkspaceException::class)
    override fun doUpdate(id: String, pattern: Comment) =
        this.doRetrieve(id).also { it.update(pattern) }

    @Throws(RepositoryException::class, WorkspaceException::class)
    override fun doDelete(id: String) =
        this.doRetrieve(id).remove()

    override fun getNodePrimaryType() = TypeNamespace.COMMENT

    override fun getDefaultRootPath() = JcrFile.ROOT_PATH

    override fun fromNode(node: Node) = JcrComment.of(node)
}
